nix/grafana/default.nix

```diff
@@ -0,0 +1,83 @@
+{ pkgs, lib, name, config, ... }:
+let
+  inherit (lib) types;
+  iniFormat = pkgs.formats.ini { };
+in
+{
+  options = {
+    description = ''
+      Configure grafana.
+    '';
+    enable = lib.mkEnableOption name;
+
+    package = lib.mkPackageOption pkgs "grafana" { };
+
+    port = lib.mkOption {
+      type = types.int;
+      description = "Which port to run grafana on.";
+      default = 3000;
+    };
+
+    dataDir = lib.mkOption {
+      type = types.str;
+      description = "Directory where grafana stores its logs and data.";
+      default = "./data/${name}";
+    };
+
+    extraConf = lib.mkOption {
+      type = iniFormat.type;
+      description = "Extra configuration for grafana.";
+      default = { };
+      example = ''
+        {
+          security.admin_user = "patato";
+          security.admin_password = "potato";
+        }
+      '';
+    };
+
+    outputs.settings = lib.mkOption {
+      type = types.deferredModule;
+      internal = true;
+      readOnly = true;
+      default = {
+        processes."${name}" =
+          let
+            grafanaConfig =
+              {
+                server = {
+                  protocol = "http";
+                  http_port = config.port;
+                  domain = "localhost";
+                };
+              } // config.extraConf;
```

Comment:
What happens if `extraConfg = { server.http_port = 8080; }`? Would it not end up removing `protocol` and `domain`? Shouldn't we do a deep merge?

Reply:
i am using `lib.recursiveUpdate` now.

i have made this same error here. i'll create a PR to fix this

- https://github.com/juspay/services-flake/blob/76beca31cc5c257f02b0fbb4433803ab4bab6a32/nix/clickhouse/default.nix#L95